usaspending_api/search/v2/views/spending_by_category_views/base_spending_by_category.py

```diff
@@ -0,0 +1,153 @@
+            logger.info(
+                f"Using experimental Elasticsearch functionality for 'spending_by_category/{self.category.name}'"
+            )
+            filter_query = QueryWithFilters.generate_transactions_elasticsearch_query(self.filters)
+            search = TransactionSearch().filter(filter_query)
+            results = self.handle_elasticsearch_logic(search)
+        else:
+            if self.subawards:
+                base_queryset = subaward_filter(self.filters)
+                self.obligation_column = "amount"
+            else:
+                base_queryset = spending_by_category_view_queryset(self.category, self.filters)
+                self.obligation_column = "generated_pragmatic_obligation"
+            results = self.handle_django_logic(base_queryset)
+
+        page_metadata = get_simple_pagination_metadata(len(results), self.pagination["limit"], self.pagination["page"])
+        response = {
+            "category": self.category.name,
+            "limit": self.pagination["limit"],
+            "page_metadata": page_metadata,
+            "results": results[: self.pagination["limit"]],
+            "messages": [get_time_period_message()],
+        }
+
+        return response
+
+    def common_db_query(self, queryset: QuerySet, django_filters: dict, django_values: list) -> QuerySet:
+        return (
+            queryset.filter(**django_filters)
+            .values(*django_values)
+            .annotate(amount=Sum(self.obligation_column))
+            .order_by("-amount")
+        )
+
+    def get_elasticsearch_results_generator(self, search: TransactionSearch, size: int = 1000) -> list:
+        """
+        The number of buckets required across the different categories varies. To solve this a generator is created
+        with the assumption that the "apply_elasticsearch_aggregations" implementation makes use of partitions.
+        This will allow for any number of results to be retrieved regardless of the number of buckets.
+        """
+        bucket_count = get_number_of_unique_terms(search, self.category.primary_field)
+        num_partitions = (bucket_count // size) + 1
+
+        for partition in range(num_partitions):
+            self.apply_elasticsearch_aggregations(search, partition, num_partitions, size)
+            response = search.handle_execute()
+            if response is None:
+                raise Exception("Breaking generator, unable to reach cluster")
+            results = self.build_elasticsearch_result(response.aggs.to_dict())
+            yield results
+
+    def handle_elasticsearch_logic(self, search: TransactionSearch) -> list:
+        results = self.get_elasticsearch_results_generator(search)
+        flat_results = list(itertools.chain.from_iterable(results))
+        return flat_results
+
+    @abstractmethod
+    def apply_elasticsearch_aggregations(
+        self, search: TransactionSearch, curr_partition: int, num_partitions: int, size: int
+    ) -> None:
```

Comment:
```suggestion
    ) -> TransactionSearch:
```
Side effects can be dangerous, particularly when working with something quirky like Elasticsearch where a non referentially-transparent function like this could be doing things a developer wouldn't expect. Having this work as a builder might make it clearer that only the TransactionSearch is being modified.